### ISSUE TITLE
Clang linters warn about typeof().

### ISSUE THREAD

**Issue by cjwelborn** [mortie/snow#25]
Specifically SublimeLinter-clang for me. It throws up:
```
implicit declaration of function 'typeof' is invalid in C99 [-Wimplicit-function-declaration]
```
..for using `asserteq()`.

I'm testing my code with both `gcc` and `clang`, because I'm making a little library. I think it can be resolved with something like this:

```c
#ifdef __clang__
    #define typeof(x) __typeof__(x)
#else
    // Might even check for __GNUC__?
    #define typeof(x) __typeof(x)
#endif
```
I know you're no stranger to [this stuff](https://mort.coffee/home/c-compiler-quirks/). I happen to
be compiling with `gcc` most of the time, but linting with `clang` because I like SublimeLinter.
That means I see a lot of these little "quirks".

**Comment by mortie** [comment]
Interesting, what version of clang are you using? It seems to work both on Clang 8 and on [CI](https://travis-ci.org/mortie/snow)'s Clang 3.8. (Specifically, https://travis-ci.org/mortie/snow/jobs/518519520#L726 compiles `test/cases/asserts.c` without warnings with clang 3.8 with `-Werror -Wall -Wextra -Wpedantic -Wincompatible-pointer-types-discards-qualifiers -Wmissing-prototypes -Wwrite-strings`, and `asserts.c` contains a lot of `asserteq` and `assertneq`)

**Comment by cjwelborn** [comment]
Okay, had to gather some info for you.

**TLDR:** Linter was using `-std=c11`, `snow` requires GNU extensions anyway (like `-std=gnu11`).

This is the minimal example (example.c):
```c
#define SNOW_ENABLED
#include "snow.h"
describe(not_a_real_test) {
    it("should not warn about asserteq") {
        asserteq(1, 1, "oops.");
    }
}
```

When compiled with `clang version 3.9.0` like the linter does (with proper include dirs for `snow.h`, and extra warning flags removed):
```
clang -std=c11 -fsyntax-only -fno-caret-diagnostics example.c
```

Results in the warning I mentioned:
```
example.c:5:9: warning: implicit declaration of function 'typeof' is invalid in C99 [-Wimplicit-function-declaration]
```

As I mentioned before, defining the `typeof()` macro for each compiler seems to work:
```c
#define SNOW_ENABLED
#ifdef __clang__
    #define typeof(x) __typeof__(x)
#else
    #define typeof(x) __typeof(x)
#endif

#include "snow.h"
describe(not_a_real_test) {
    it("should not warn about asserteq") {
        asserteq(1, 1, "oops.");
    }
}
```

..same flags, same command, no warnings from `clang`.

However, removing `-std=c11` also removes the warning (just like on your [travis build](https://travis-ci.org/mortie/snow/jobs/518519520#L726)).

I added the `-std=c11` because I use the linter on many different projects, that may or may not use GNU extensions. For projects that use the extensions, I add the `_GNU_SOURCE` definition, which the linter
doesn't pick up on. It could if I wanted it to, but that would be like using `-std=gnu11`.

Looking at [clang's language compatibility page](https://clang.llvm.org/compatibility.html#inline), it says 

> By default, Clang builds C code in GNU C11 mode, ...

..which would be `-std=gnu11` I guess. Using `-std=gnu11` also removes the warnings.

For now, I don't know what will happen if I decide to lint all of my projects using `-std=gnu11`. I'll have to try it and see. I know users of `snow.h` are expected to use the gnu extensions. The users of my little library that I'm working on would be expected to use the gnu extensions too (not because of `snow`). I was thinking about some weird case where the tests use the extensions, but the actual application doesn't. I don't know if there's a case like that in the real world, so I'll just fix my linter.

At least this will be documentation if someone else comes across this little SublimeLinter "gotcha".

**Comment by cjwelborn** [comment]
After playing around with `-std=gnu11`, if you compile with that "macro fix" above AND `-std=gnu11` it results in another warning (caused by the macro, not `snow`):
```
example2.c:3:13: warning: keyword is hidden by macro definition [-Wkeyword-macro]
``

...so, it's either `-std=c11` with the macro, or `-std=gnu11` without it. Not both.
